infrastructure/src/test/java/org/corfudb/infrastructure/management/failuredetector/EpochHandlerTest.java

```diff
@@ -0,0 +1,181 @@
+                .join();
+
+        assertTrue(latestLayout.isPresent());
+    }
+
+    @Test
+    void fetchLatestLayoutThreeNodesOneDown() {
+        final int highestEpoch = 33;
+
+        CompletableFuture<Layout> cf0 = CompletableFuture.completedFuture(buildSimpleLayout());
+        CompletableFuture<Layout> cf1 = CompletableFuture.completedFuture(buildSimpleLayout(highestEpoch));
+        CompletableFuture<Layout> cf2 = new CompletableFuture<>();
+        cf2.completeExceptionally(new WrongEpochException(123));
+
+        HashMap<String, CompletableFuture<Layout>> futureLayouts = new HashMap<>();
+        futureLayouts.put(NodeNames.A, cf0);
+        futureLayouts.put(NodeNames.B, cf1);
+        futureLayouts.put(NodeNames.C, cf2);
+
+        Optional<Layout> latestLayout = buildHandler()
+                .fetchLatestLayout(futureLayouts)
+                .join();
+
+        assertTrue(latestLayout.isPresent());
+        assertEquals(highestEpoch, latestLayout.get().getEpoch());
+    }
+
+    @Test
+    void updateTrailingLayoutServersCommitLayoutExecution() {
+        EpochHandler handlerSpy = Mockito.spy(buildHandler());
+
+        HashMap<String, CompletableFuture<Layout>> requests = new HashMap<>();
+        CompletableFuture<Layout> request1 = new CompletableFuture<>();
+        request1.complete(buildSimpleLayout(2));
+        requests.put(NodeNames.A, request1);
+
+        CompletableFuture<Boolean> asyncCommit = CompletableFuture.completedFuture(true);
+        doReturn(asyncCommit).when(handlerSpy).commitLayout(isA(Layout.class), anyString());
+
+        Layout layout1 = buildSimpleLayout();
+        handlerSpy.updateTrailingLayoutServers(layout1, requests);
+
+        verify(handlerSpy, times(1)).commitLayout(any(Layout.class), anyString());
+    }
+
+    @Test
+    void updateTrailingLayoutServersFailedConnections() {
+        EpochHandler handlerSpy = Mockito.spy(buildHandler());
+
+        CompletableFuture<Layout> request1 = new CompletableFuture<>();
+        CompletableFuture<Layout> request2 = new CompletableFuture<>();
+        CompletableFuture<Layout> request3 = new CompletableFuture<>();
+        request1.complete(buildSimpleLayout(2));
+        request2.completeExceptionally(new TimeoutException("Connection error"));
+        request3.completeExceptionally(new TimeoutException("Connection error"));
+
+        HashMap<String, CompletableFuture<Layout>> requests = new HashMap<>();
+        requests.put(NodeNames.A, request1);
+        requests.put(NodeNames.B, request2);
+        requests.put(NodeNames.C, request3);
+
+        CompletableFuture<Boolean> asyncCommit = CompletableFuture.completedFuture(true);
+        doReturn(asyncCommit).when(handlerSpy).commitLayout(isA(Layout.class), anyString());
+
+        Layout layout1 = buildSimpleLayout();
+        handlerSpy.updateTrailingLayoutServers(layout1, requests).join();
+
+        verify(handlerSpy, times(3)).commitLayout(any(Layout.class), anyString());
```

Comment:
Since we check that `commitLayout` has been executed, 
 - I think there is no need for extra checks, especially when those checks are internal
 - I wouldn't like to expose too much of the internal state/execution, it doesn't seem reasonable

Overall: I think, it would be great not to use mocks at all, or at least minimize using them for the sake of simplicity